src/spikeinterface/extractors/nwbextractors.py

```diff
@@ -798,201 +915,183 @@ def get_traces(self, start_frame, end_frame, channel_indices):
+        elif hasattr(self, "_nwbfile"):
+            io = self._nwbfile.get_read_io()
+            if io is not None:
+                io.close()
+
+    def extract_info_pynwb(
+        self, unit_table_path: str = None, samples_for_rate_estimation: int = 1000, cache: bool = False
     ):
-        try:
-            from pynwb import NWBHDF5IO, NWBFile
-            from pynwb.ecephys import ElectrodeGroup
-        except ImportError:
-            raise ImportError(self.installation_mesg)
-
-        self.stream_mode = stream_mode
-        self.stream_cache_path = stream_cache_path
-        self._electrical_series_name = electrical_series_name
-
-        self.file_path = file_path
         self._nwbfile = read_nwbfile(
-            file_path=file_path, stream_mode=stream_mode, cache=cache, stream_cache_path=stream_cache_path
+            backend=self.backend,
+            file_path=self.file_path,
+            stream_mode=self.stream_mode,
+            cache=cache,
+            stream_cache_path=self.stream_cache_path,
+            storage_options=self.storage_options,
         )
 
         timestamps = None
-        self.t_start = t_start
-        if sampling_frequency is None:
+        if self.internal_sampling_frequency is None:
             # defines the electrical series from where the sorting came from
             # important to know the sampling_frequency
-            self.electrical_series = retrieve_electrical_series(self._nwbfile, self._electrical_series_name)
+            self.electrical_series = retrieve_electrical_series(self._nwbfile, self.electrical_series_path)
             # get rate
             if self.electrical_series.rate is not None:
-                sampling_frequency = self.electrical_series.rate
+                self.internal_sampling_frequency = self.electrical_series.rate
                 self.t_start = self.electrical_series.starting_time
             else:
                 if hasattr(self.electrical_series, "timestamps"):
                     if self.electrical_series.timestamps is not None:
                         timestamps = self.electrical_series.timestamps
-                        sampling_frequency = 1 / np.median(np.diff(timestamps[samples_for_rate_estimation]))
+                        self.internal_sampling_frequency = 1 / np.median(
+                            np.diff(timestamps[:samples_for_rate_estimation])
+                        )
                         self.t_start = timestamps[0]
         assert (
-            sampling_frequency is not None
+            self.internal_sampling_frequency is not None
         ), "Couldn't load sampling frequency. Please provide it with the 'sampling_frequency' argument"
         assert (
             self.t_start is not None
         ), "Couldn't load a starting time for the sorting. Please provide it with the 't_start' argument"
 
-        units_table = self._nwbfile.units
+        units_table = retrieve_unit_table(self._nwbfile, unit_table_path=unit_table_path)
```

Comment:
The pynwb double route still performs the search even if the units table path is passed I think.